src/integration/integration.service.ts

```diff
@@ -0,0 +1,352 @@
+      });
+    }
+
+    const { promotions, availablePromotions } =
+      await this.voucherifyService.getPromotions(cart, uniqCoupons);
+
+    if (!uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'No coupons applied, skipping voucherify call',
+      });
+
+      return {
+        availablePromotions,
+      };
+    }
+
+    const deletedCoupons = uniqCoupons.filter(
+      (coupon) => coupon.status === 'DELETED',
+    );
+
+    deletedCoupons
+      .filter((coupon) => coupon.type !== 'promotion_tier')
+      .map((coupon) =>
+        this.voucherifyConnectorService.releaseValidationSession(
+          coupon.code,
+          sessionKey,
+        ),
+      );
+
+    if (deletedCoupons.length === uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'Deleting coupons only, skipping voucherify call',
+      });
+
+      return {
+        availablePromotions,
+      };
+    }
+
+    this.logger.debug({
+      msg: 'Attempt to apply coupons',
+      coupons: uniqCoupons,
+      id,
+      customerId,
+      anonymousId,
+    });
+
+    uniqCoupons = filterCouponsByLimit(
+      uniqCoupons,
+      getCouponsLimit(
+        this.configService.get<number>('COMMERCE_TOOLS_COUPONS_LIMIT'),
+      ),
+    );
+
+    let validatedCoupons: ValidationValidateStackableResponse =
+      await this.voucherifyConnectorService.validateStackableVouchers(
+        buildValidationsValidateStackableParamsForVoucherify(
+          uniqCoupons.filter((coupon) => coupon.status != 'DELETED'),
+          cart,
+          this.productMapper.mapLineItems(cart.lineItems),
+          sessionKey,
+        ),
+      );
+
+    const productsToAdd = await this.commercetoolsService.getProductsToAdd(
+      validatedCoupons,
+      this.commercetoolsService.getPriceSelectorFromCart(cart),
+    );
+
+    const productsToAddWithIncorrectPrice = productsToAdd.filter(
+      (product) => product.discount_difference,
+    );
+
+    if (productsToAddWithIncorrectPrice.length) {
+      const items = await this.getItemsWithCorrectedPrices(
+        validatedCoupons,
+        productsToAddWithIncorrectPrice,
+      );
+      validatedCoupons =
+        await this.voucherifyConnectorService.validateStackableVouchers(
+          buildValidationsValidateStackableParamsForVoucherify(
+            uniqCoupons.filter((coupon) => coupon.status != 'DELETED'),
+            cart,
+            items,
+            sessionKey,
+          ),
+        );
+    }
+
+    if (promotions.length) {
+      this.voucherifyService.setBannerOnValidatedPromotions(
+        validatedCoupons,
+        promotions,
+      );
+    }
+
+    this.logger.debug({
+      msg: 'Validated coupons',
+      validatedCoupons,
+      availablePromotions,
+      productsToAdd,
+    });
+
+    return {
+      validatedCoupons,
+      availablePromotions,
+      productsToAdd,
+    };
+  }
+
+  private async getItemsWithCorrectedPrices(
+    validatedCoupons: ValidationValidateStackableResponse,
```

Comment:
we don't need to provide the whole `validatedCoupons`, we can narrow this param down to `validatedCoupons.order.items`